Subscriber middleware

EtherZa opened the pull request:
This is a partial implementation of the feature requested in #218 (limited to subscriber middleware)

The change modifies `HandlerInvoker` to add middleware to the execution pipeline. 

I acknowledge that both the PR and implementation are unsolicited, and am happy to modify/withdraw as required.

Example usage:
```cs
services.AddAWSMessageBus(builder =>
{
    builder.AddMiddleware<LoggingMiddleware>();  // Singleton by default
    ...
}

public class LoggingMiddleware(ILogger<LoggingMiddleware> Logger) : IMiddleware
{
    public Task<MessageProcessStatus> InvokeAsync<T>(MessageEnvelope<T> messageEnvelope, RequestDelegate next, CancellationToken token = default)
    {
        var stopwatch = Stopwatch.StartNew();
        try
        {
            return next();
        }
        finally
        {
            stopwatch.Stop();
            Logger.LogInformation("Processed message with ID: {MessageId} and Type: {MessageType} in {Elapsed}",
                messageEnvelope.Id,
                messageEnvelope.MessageTypeIdentifier,
                stopwatch.Elapsed);
        }
    }
}
```


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

EtherZa committed:
Subscriber middleware

EtherZa committed:
Subscriber middleware

EtherZa commented:
Typos and documentation added as per copilot. Squashed and rebased on `dev`.

EtherZa committed:
Message error handler

EtherZa committed:
Message error handler

jdnickell commented:
Just wanted to express interest in this feature specifically for the ability to use custom retry policies. Are there any concerns with this approach or anything in particular blocking this review besides availability?

GarrettBeatty commented:
hi sorry for the delay - taking a look at this again

GarrettBeatty committed:
Add AddMessageErrorHandler to IMessageBusBuilder and add XML documentation to IMessageErrorHandler

- Added AddMessageErrorHandler<T> method declaration to IMessageBusBuilder interface for consistency with AddMiddleware<T>

- Added XML summary documentation to IMessageErrorHandler interface and MessageErrorHandlerResponse enum

- Fixed missing space before colon in generic constraint (where T: -> where T :)

GarrettBeatty committed:
Add AddMessageErrorHandler to IMessageBusBuilder and add XML documentation to IMessageErrorHandler

- Added AddMessageErrorHandler<T> method declaration to IMessageBusBuilder interface for consistency with AddMiddleware<T>

- Added XML summary documentation to IMessageErrorHandler interface and MessageErrorHandlerResponse enum

- Fixed missing space before colon in generic constraint (where T: -> where T :)

GarrettBeatty committed:
Add documentation for subscriber middleware and message error handler

- Added Subscriber Middleware section to main README with IMiddleware example and registration

- Added Message Error Handling section to main README with IMessageErrorHandler example and MessageErrorHandlerResponse descriptions

- Updated SubscriberService sample README to document middleware and project structure

GarrettBeatty committed:
Add documentation for subscriber middleware and message error handler

- Added Subscriber Middleware section to main README with IMiddleware example and registration

- Added Message Error Handling section to main README with IMessageErrorHandler example and MessageErrorHandlerResponse descriptions

- Updated SubscriberService sample README to document middleware and project structure

GarrettBeatty committed:
Add additional unit tests for middleware coverage gaps

- Middleware_WithMultipleMessageTypes_RoutesCorrectly: verifies delegate dispatch with ChatMessage and AddressInfo through shared middleware

- Middleware_ShortCircuit_DoesNotExecuteDownstream: verifies middleware can return without calling next() and downstream is skipped

- Middleware_CancellationToken_IsPropagatedThroughPipeline: verifies cancellation token reaches middleware

- Middleware_SingletonVsScoped_LifetimeBehavior: verifies scoped middleware creates new instances per invocation while singleton reuses

GarrettBeatty committed:
Add additional unit tests for middleware coverage gaps

- Middleware_WithMultipleMessageTypes_RoutesCorrectly: verifies delegate dispatch with ChatMessage and AddressInfo through shared middleware

- Middleware_ShortCircuit_DoesNotExecuteDownstream: verifies middleware can return without calling next() and downstream is skipped

- Middleware_CancellationToken_IsPropagatedThroughPipeline: verifies cancellation token reaches middleware

- Middleware_SingletonVsScoped_LifetimeBehavior: verifies scoped middleware creates new instances per invocation while singleton reuses

GarrettBeatty commented on src/AWS.Messaging/Services/HandlerInvoker.cs:
```diff
@@ -16,31 +15,35 @@ public class HandlerInvoker : IHandlerInvoker
-    /// Caches the <see cref="MethodInfo"/> of the <see cref="IMessageHandler{T}.HandleAsync(MessageEnvelope{T}, CancellationToken)"/>
-    /// method that will be invoked with the message envelope for each handler
-    /// </summary>
-    private readonly ConcurrentDictionary<Type, MethodInfo?> _handlerMethods = new();
```
this change was took a while for me to understand but i think it makes sense. here is my understanding for others (ai summarized)


Okay, let me walk through the complete lifecycle with a concrete example.

## Your application code:

```csharp
services.AddAWSMessageBus(builder =>
{
    builder.AddMessageHandler<ChatMessageHandler, ChatMessage>("chatMessage");
    builder.AddMiddleware<LoggingMiddleware>();
});
```

---

## Phase 1: Startup (happens once)

When `builder.AddMessageHandler<ChatMessageHandler, ChatMessage>("chatMessage")` runs, it calls:

```csharp
// In SubscriberMapping.cs
public static SubscriberMapping Create<THandler, TMessage>(string? messageTypeIdentifier = null)
```

Here, `THandler = ChatMessageHandler` and `TMessage = ChatMessage`. Inside this method, a **static local function** is defined:

```csharp
static Task<MessageProcessStatus> handlerInvoker(
    HandlerInvoker invoker, 
    MessageEnvelope messageEnvelope,    // <-- this is the BASE type, not generic
    SubscriberMapping subscriberMapping, 
    CancellationToken token)
{
    // Cast the base envelope to the specific generic type, then call the generic method
    return invoker.InvokeAsync((MessageEnvelope<ChatMessage>)messageEnvelope, subscriberMapping, token);
}
```

Because this function is inside `Create<..., ChatMessage>()`, the compiler replaces `TMessage` with `ChatMessage`. This function is now a **regular compiled function** that does two things:
1. Casts `MessageEnvelope` → `MessageEnvelope<ChatMessage>`
2. Calls `invoker.InvokeAsync<ChatMessage>(...)`

This function reference is stored as a delegate on the new `SubscriberMapping`:

```csharp
var mapping = new SubscriberMapping(
    handlerType: typeof(ChatMessageHandler),
    messageType: typeof(ChatMessage),
    envelopeFactory: ...,
    handlerInvoker: handlerInvoker,    // <-- the function above, stored as a delegate
    messageTypeIdentifier: "chatMessage"
);
```

The mapping is added to the configuration's list. **Startup is done.** The delegate exists and is ready to be called.

---

## Phase 2: A message arrives from SQS (happens per message)

SQS delivers a message. The framework:
1. Reads the message body
2. Finds the type identifier `"chatMessage"` in the CloudEvent envelope
3. Looks up the `SubscriberMapping` with `messageTypeIdentifier == "chatMessage"`
4. Deserializes the data into a `MessageEnvelope<ChatMessage>` (but holds it as the base `MessageEnvelope`)

Now it calls:

```csharp
handlerInvoker.InvokeAsync(envelope, subscriberMapping, token);
```

### Step 1: Non-generic entry point

```csharp
// In HandlerInvoker.cs
public Task<MessageProcessStatus> InvokeAsync(
    MessageEnvelope messageEnvelope,      // base type
    SubscriberMapping subscriberMapping, 
    CancellationToken token)
{
    // Call the delegate stored on the mapping
    return subscriberMapping.HandlerInvoker(this, messageEnvelope, subscriberMapping, token);
}
```

This is just calling the delegate we stored at startup. The delegate is literally just:
```csharp
return invoker.InvokeAsync((MessageEnvelope<ChatMessage>)messageEnvelope, subscriberMapping, token);
```

So it casts the envelope and calls...

### Step 2: Generic method (T is now ChatMessage)

```csharp
// In HandlerInvoker.cs
public async Task<MessageProcessStatus> InvokeAsync<T>(
    MessageEnvelope<T> messageEnvelope,    // <-- T = ChatMessage, strongly typed!
    SubscriberMapping subscriberMapping, 
    CancellationToken token)
{
    // Create a new DI scope for this message
    await using var scope = _serviceProvider.CreateAsyncScope();

    // Resolve the handler - cast is safe because we know THandler implements IMessageHandler<ChatMessage>
    IMessageHandler<ChatMessage> handler = (IMessageHandler<ChatMessage>)
        scope.ServiceProvider.GetRequiredService(typeof(ChatMessageHandler));

    // Resolve all registered middleware
    List<IMiddleware> middlewares = [loggingMiddleware];  // resolved from DI

    // Build and execute the pipeline
    return await ExecutePipelineAsync(messageEnvelope, middlewares, handler, token);
}
```

### Step 3: Execute the pipeline

```csharp
private static async Task<MessageProcessStatus> ExecutePipelineAsync<T>(
    MessageEnvelope<T> messageEnvelope, 
    List<IMiddleware> middlewares, 
    IMessageHandler<T> handler, 
    CancellationToken token)
{
    // Start with the innermost layer: the handler itself
    RequestDelegate next = () => handler.HandleAsync(messageEnvelope, token);
    // next = () => chatMessageHandler.HandleAsync(envelope, token)

    // Wrap with middleware from inside out (iterate backwards)
    for (var i = middlewares.Count - 1; i >= 0; i--)
    {
        var capturedNext = next;
        var middleware = middlewares[i];
        next = () => middleware.InvokeAsync(messageEnvelope, capturedNext, token);
        // next = () => loggingMiddleware.InvokeAsync<ChatMessage>(envelope, capturedNext, token)
    }

    // Execute: calls LoggingMiddleware, which calls next(), which calls ChatMessageHandler
    return await next().ConfigureAwait(false);
}
```

### Step 4: Execution flow

```
next() is called
  → LoggingMiddleware.InvokeAsync<ChatMessage>(envelope, next, token)
    → logger.LogInformation("Processing message...")
    → next() is called
      → ChatMessageHandler.HandleAsync(envelope, token)
      → returns MessageProcessStatus.Success()
    → logger.LogInformation("Done processing")
    → returns MessageProcessStatus.Success()
  → returns MessageProcessStatus.Success()
```

---

## The entire flow in one picture:

```
SQS Message arrives
  ↓
Framework finds SubscriberMapping for "chatMessage"
  ↓
Calls InvokeAsync(baseEnvelope, mapping, token)         [non-generic, HandlerInvoker.cs]
  ↓
Calls mapping.HandlerInvoker delegate                    [the delegate stored at startup]
  ↓
Delegate casts: (MessageEnvelope<ChatMessage>)envelope   [one cast, no reflection]
  ↓
Calls InvokeAsync<ChatMessage>(typedEnvelope, ...)       [generic, T=ChatMessage known]
  ↓
Creates DI scope, resolves handler + middleware
  ↓
Builds pipeline: LoggingMiddleware → ChatMessageHandler
  ↓
Executes pipeline: Logging wraps around Handler
  ↓
Returns MessageProcessStatus.Success()
```

The delegate is the one-line bridge that turns a base `MessageEnvelope` into a `MessageEnvelope<ChatMessage>`. Everything downstream is strongly typed and directly called - no reflection anywhere in the per-message path.

GarrettBeatty replied:
https://github.com/aws/aws-dotnet-messaging/pull/330/ this is similar to what this pr does

GarrettBeatty commented:
https://github.com/aws/aws-dotnet-messaging/actions/runs/25005191911 ci run

philasmar commented on src/AWS.Messaging/Configuration/MessageBusBuilder.cs:
```diff
@@ -283,6 +304,29 @@ MessageEnvelope envelopeFactory()
+            var mappingParam = Expression.Parameter(typeof(SubscriberMapping), "mapping");
+            var tokenParam = Expression.Parameter(typeof(CancellationToken), "token");
+
+            // invoker.InvokeAsync<T>( (MessageEnvelope<T>) envelope, mapping, token )
```
nit: remove comment

EtherZa replied:
Removed.

philasmar commented on src/AWS.Messaging/IMiddleware.cs:
```diff
@@ -0,0 +1,30 @@
+/// <summary>
+/// This interface is implemented by the users of this library for each layer of middleware that should be processed.
+/// </summary>
+public interface IMiddleware
```
nit: not a fan of using a very generic name. Maybe `IMessageMiddleware`?

EtherZa replied:
That makes sense -especially since there is potential for middleware support to be added to the publishing pipeline later. Would `IHandlerMiddleware` be a better fit here? That would also leave room for something like `IPublishMiddleware` in the future.

It may also be worth updating the registration API to reflect that, for example:
`IMessageBusBuilder.AddMiddleware<T>()` → `IMessageBusBuilder.AddHandlerMiddleware<T>()`

GarrettBeatty replied:
I like `IHandlerMiddleware`

EtherZa replied:
Renamed to `IHandlerMiddleware`

GarrettBeatty replied:
>It may also be worth updating the registration API to reflect that, for example:
IMessageBusBuilder.AddMiddleware<T>() → IMessageBusBuilder.AddHandlerMiddleware<T>()

I think we should also do this too?

EtherZa replied:
Done.

EtherZa committed:
Rename IMiddleware to IHandlerMiddleware
Remove code comment

EtherZa commented:
Hi @GarrettBeatty / @philasmar  . Thanks for the review.

Following the comments, I have:
* rebased on `dev` (required the `SubscriberMiddleware` property to be added to `SingleTypeMessageConfiguration`)
* removed the comment from `MessageBusBuilder`
* renamed `IMiddleware` to `IHandlerMiddleware`

No other changes have been made.

EtherZa committed:
Rename IMessageBusBuilder.AddMiddleware(...) to .AddHandlerMiddleware(...)
